java/src/com/ibm/streamsx/topology/kafka/KafkaStreams.java

```diff
@@ -0,0 +1,284 @@
+    }
+    
+    private static void addPropertiesFile(TopologyElement te) {
+        // TODO create tmp empty properties file "etc/emptyKafkaAdapterProperties"
+       /*
+         File tmpDir = mkTmpDir
+         File etcDir = mkDir(tmpDir, "etc");
+         File props = mkFile(tmpDir, PROPS_FILE);
+         props.createNewFile();
+         
+        te.addFileDependency(props.getAbsolutePath());
+        
+        when will it get removed?  if instead Topology makes the copy...???
+         */
+    }
+
+    /**
+     * Convert an {@link SPLStream} with schema {@link KafkaSchemas.KAFKA}
+     * to a TStream&lt;{@link KafkaMessageImpl}>.
+     * The returned stream will contain a {@code KafkaTuple} tuple for
+     * each tuple {@code T} on {@code stream}.
+     * A runtime error will occur if the schema of {@code stream} doesn't
+     * have the attributes defined by {@code KafkaSchemas.KAFKA}.
+     * @param stream Stream to be converted to a TStream&lt;KafkaTuple>.
+     * @return Stream of {@code KafkaTuple} tuples from {@code stream}.
+     */
+    private static TStream<KafkaMessage> toKafkaTupleStream(SPLStream stream) {
+
+        return stream.convert(new Function<Tuple, KafkaMessage>() {
+            private static final long serialVersionUID = 1L;
+
+            @Override
+            public KafkaMessage apply(Tuple tuple) {
+                return new KafkaMessageImpl(fromSplValue(tuple.getString("message")),
+                                    fromSplValue(tuple.getString("key")));
+            }
+            
+            private String fromSplValue(String s) {
+                // SPL doesn't allow null value.  For our use,
+                // assume an empty string meant null.
+                return (s==null || s.isEmpty()) ? null : s;
+            }
+        }, KafkaMessage.class);
+    }
+
+    /**
+     * Send a stream of messages to a Kafka Cluster.
+     * This is an adapter to a Kafka KafkaProducer.
+     * <p>
+     * N.B., A topology that includes this will not support
+     * {@code StreamsContext.EMBEDDED}.
+     *  
+     * @param kafkaProducerProperties see the Kafka documentation.
+     *          Minimal configuration typically includes:
+     * <ul>
+     * <li><code>metadata.broker.list</code></li>
+     * <li><code>serializer.class</code></li>
+     * <li><code>request.required.acks</code></li>
+     * </ul>
+     * @param topic optional Kafka topic to publish to.
+     *          Whens specified, a {@link KafkaMessageImpl#getKafkaTopic()} value is ignored.
+     *          If null, {@code toKafkaTupleFunc} must generate tuples that have a topic.
+     * @param stream the stream to publish to Kafka
+     * @param toKafkaMessageFunc function to generate a {@code KafkaTuple} from a tuple {@code T}.
+     *          If {@code topic} is null, the generated {@code KafkaTuple} <b>must</b> contain a {@code topic}.
+     */
+    public static <T> void sink(TopologyElement te,
```

Comment:
Good question.  Originally I had "producer" and "consumer" but found it rather confusing (in usage in the test code).  I switched to "sink" and "source" to make it less confusing.  "sink" and "source" are the Java API methods that would have been called/used by a user had I been able to instead offer up the functionality as a Java API "Supplier" and a "Consumer" implementation respectively.  And of course, the "Consumer" implementation would really be a kafka-producer. :-)
Happy to change it to whatever you want.  Just say the word.